### PR TITLE
Abstract common util functions into sensible module

### PR DESCRIPTION
**Pull request by aiolos404**
Refactor the shared non-coin functions into utils/validate.ts and hash.ts 

Pull request aims to solve code challenge: https://github.com/BitGo/bitgo-account-lib/issues/57#issue-561945458

is-related functions are moved to validate.ts and called as Validator module in other files
HashType interface and hashTypes model are moved to hash.ts for decoupling purpose

unit test refined

### CODE REVIEW

File: src/utils/hash.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    prefix: new Buffer([69, 220, 169]),
+    byteLength: 15,
+  },
+  /* 4 bytes long */
+  // chain id
+  id: {
+    prefix: new Buffer([153, 103]),
+    byteLength: 4,
+  },
+};
```
[the remark this answers is not in the document]
Comment (aiolos404) [reply src/utils/hash.ts]: I checked these has types are for Base58Check. As far as I know, Base58Check is used for bitcoin address encoding, not for Tezos specifically. That the reason why I proposed to move them into a common util file

File: src/utils/hash.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
+export interface HashType {
+  prefix: Buffer;
+  byteLength: number;
+}
```
[the remark this answers is not in the document]
Comment (aiolos404) [reply src/utils/hash.ts]: I agree

File: src/utils/validate.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  // Check prefix
+  if (!hashPrefix.equals(Buffer.from(hashType.prefix))) {
+    return false;
+  }
+
+  // Check length
+  const hashLength = decodedHash.length - hashPrefix.length;
+  return hashLength === hashType.byteLength;
+}
```
[the remark this answers is not in the document]
Comment (aiolos404) [reply src/utils/validate.ts]: I was planning to create an interface for a "validator" which will have all abstract `isValid*() `methods in a common util file. But the implementation of the interface should be in util files for each specific coin implementation. It does not make sense to me just move interface to a new file. 
I think I should propose to implement `isValid*()` methods with a coin type argument. That might decrease the LOC.

File: src/utils/hash.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (aiolos404) [reply src/utils/hash.ts]: Done in commit 75c93d5

File: src/utils/hash.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (CapnMigraine) [reply src/utils/hash.ts:136]: That's true, although these specific types and prefixes are used by Tezos only, the encoding is common to other coins.
But we will not verify hashes with this method for other coins. We already use methods like `isValidAddress` or `isValidPub`

File: src/utils/validate.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (CapnMigraine) [reply src/utils/validate.ts:38]: The interface approach would be ideal, but we also need to implement those methods for Tron, which is the only other coin supported by the library right now. I think that's out of the scope of this PR

File: src/utils/hash.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (aiolos404) [reply src/utils/hash.ts]: OK, I got it. Thanks for clarification. I moved hashTypes back to `xtz/utils.ts` in commit 36cd8d2

File: src/utils/validate.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (aiolos404) [reply src/utils/validate.ts:37]: Thanks for clarification. I moved `isValid*()` function back to `xtz/utils.ts` in commit 36cd8d2

**Comment by aiolos404** [comment]
Thanks for the review. Really learned a lot in this PR.
One last question, I read the rule of the commits and signed most recent commit. Shall I sign all the previous commits and then force push?
Thanks

**Comment by aiolos404** [comment]
Been waiting for the review too long. Thank you for the reviewing process.

**Pull request closed without merging**